// testcafe/lib/proxyless/native-dialogs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const api_base_1 = __importDefault(require("../api-base"));
class NativeDialogsAPI extends api_base_1.default {
    constructor(browserId, client) {
        super(browserId, client);
        this._nativeDialogs = {};
    }
    async init() {
        this._client.Page.on('javascriptDialogOpening', async ({ type, message, url }) => {
            await this._client.Page.handleJavaScriptDialog({ accept: true });
            this._nativeDialogs[this._testRun.id] = this._nativeDialogs[this._testRun.id] || [];
            this._nativeDialogs[this._testRun.id].unshift({
                text: message,
                type,
                url,
            });
        });
    }
    async getNativeDialogHistory() {
        const nativeDialogs = this._nativeDialogs[this._testRun.id] || [];
        return nativeDialogs;
    }
    async fixMissingBeforeUnloadHandling() {
        // NOTE: do fake mouse event to make cdp handle the `beforeunload` event.
        // the `beforeunload` event is not handled without this click.
        await this._client.Input.dispatchMouseEvent({
            type: 'mousePressed',
            button: 'left',
            x: -1,
            y: -1,
        });
    }
}
exports.default = NativeDialogsAPI;
module.exports = exports.default;
