// testcafe/lib/proxyless/console-messages/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const formatter_1 = __importDefault(require("./formatter"));
const api_base_1 = __importDefault(require("../api-base"));
const types_1 = require("./types");
class ConsoleMessagesAPI extends api_base_1.default {
    constructor(browserId, client) {
        super(browserId, client);
        this._consoleMessages = {};
    }
    _getBrowserConsoleMessages(url, type) {
        if (!this._consoleMessages[this._testRun.id])
            this._consoleMessages[this._testRun.id] = {};
        const testRunMessages = this._consoleMessages[this._testRun.id];
        if (!testRunMessages[url]) {
            testRunMessages[url] = {
                error: [],
                warning: [],
                log: [],
                info: [],
            };
        }
        return testRunMessages[url][type];
    }
    async init() {
        this._client.Runtime.on('consoleAPICalled', async ({ type, args }) => {
            if (!types_1.CONSOLE_MESSAGES_ENTRY_KEYS.includes(type))
                return;
            const { host } = await this._getCurrentUrl();
            const messages = this._getBrowserConsoleMessages(host, type);
            messages.push(formatter_1.default.format(args));
        });
    }
    static _createBrowserConsoleMessages(info, log, error, warn) {
        return {
            info,
            log,
            error,
            warn,
        };
    }
    async getBrowserConsoleMessages() {
        const { host } = await this._getCurrentUrl();
        const testRunMessages = this._consoleMessages[this._testRun.id];
        if (!testRunMessages || !testRunMessages[host])
            return ConsoleMessagesAPI._createBrowserConsoleMessages([], [], [], []);
        const messages = JSON.parse(JSON.stringify(testRunMessages[host]));
        return ConsoleMessagesAPI._createBrowserConsoleMessages(messages.info, messages.log, messages.error, messages.warning);
    }
}
exports.default = ConsoleMessagesAPI;
module.exports = exports.default;
