// testcafe/lib/compiler/test-file/formats/es-next/compiler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const load_libs_1 = __importDefault(require("../../../babel/load-libs"));
const api_based_1 = __importDefault(require("../../api-based"));
const is_flow_code_1 = __importDefault(require("./is-flow-code"));
const get_base_babel_options_1 = __importDefault(require("../../../babel/get-base-babel-options"));
const disable_v8_optimization_note_1 = __importDefault(require("../../disable-v8-optimization-note"));
const extensions_1 = __importDefault(require("../extensions"));
//NOTE: The semicolon ; prevents the declaration from being bound with eval
const DISABLE_V8_OPTIMIZATION_CODE = `;/*${disable_v8_optimization_note_1.default}*/
eval("");
`;
class ESNextTestFileCompiler extends api_based_1.default {
    static getBabelOptions(filename, code, { isCompilerServiceMode, experimentalEsm } = {}) {
        const { presetStage2, presetFlow, transformRuntime, presetEnvForTestCode, presetReact, moduleResolver, proposalPrivateMethods, proposalClassProperties, } = (0, load_libs_1.default)({ isCompilerServiceMode, experimentalEsm });
        const opts = Object.assign({}, get_base_babel_options_1.default, {
            presets: [presetStage2, presetEnvForTestCode, presetReact],
            plugins: [transformRuntime, moduleResolver, proposalPrivateMethods, proposalClassProperties],
            sourceMaps: 'inline',
            filename,
        });
        if ((0, is_flow_code_1.default)(code))
            opts.presets.push(presetFlow);
        return opts;
    }
    _compileCode(code, filename) {
        const { babel } = (0, load_libs_1.default)(this);
        if (this.cache[filename])
            return this.cache[filename];
        if (this.isCompilerServiceMode || this.experimentalEsm)
            code += DISABLE_V8_OPTIMIZATION_CODE;
        const opts = ESNextTestFileCompiler.getBabelOptions(filename, code, this);
        const compiled = babel.transform(code, opts);
        this.cache[filename] = compiled.code;
        return compiled.code;
    }
    _getRequireCompilers() {
        const requireCompilers = {
            [extensions_1.default.js]: (code, filename) => this._compileCode(code, filename),
            [extensions_1.default.jsx]: (code, filename) => this._compileCode(code, filename),
            [extensions_1.default.cjs]: (code, filename) => this._compileCode(code, filename),
        };
        if (this.experimentalEsm)
            requireCompilers[extensions_1.default.mjs] = (code, filename) => this._compileCode(code, filename);
        return requireCompilers;
    }
    get canCompileInEsm() {
        return true;
    }
    getSupportedExtension() {
        const supportedExtensions = [extensions_1.default.js, extensions_1.default.jsx, extensions_1.default.cjs];
        if (this.experimentalEsm)
            supportedExtensions.push(extensions_1.default.mjs);
        return supportedExtensions;
    }
}
exports.default = ESNextTestFileCompiler;
module.exports = exports.default;
