// testcafe/lib/proxyless/js-error-handling/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../../shared/errors");
const api_base_1 = __importDefault(require("../api-base"));
const test_run_1 = require("../../errors/test-run");
const UNCAUGHT_PROMISE_REJECTION = 'Uncaught (in promise)';
const UNCAUGHT_JS_ERROR = 'Uncaught';
const NO_STACK_TRACE_AVAILABLE_MESSAGE = 'No stack trace available';
class JSErrorHandlingAPI extends api_base_1.default {
    constructor(browserId, client) {
        super(browserId, client);
        this.command = null;
    }
    async init() {
        this._client.Runtime.on('exceptionThrown', async ({ exceptionDetails }) => {
            this._addTestRunEvents();
            const errorMessage = this._getErrorMessage(exceptionDetails);
            const errFilter = this._getErrFilter(exceptionDetails, errorMessage);
            try {
                if (!this._getShouldSkipJsErrors(errFilter))
                    this._testRun.addProxylessJSError(new errors_1.UncaughtErrorOnPage(errorMessage, exceptionDetails.url));
            }
            catch (err) {
                this._testRun.addProxylessJSError(new test_run_1.UncaughtErrorInTestCode(err));
            }
        });
    }
    setErrorHandler(command) {
        this.command = command;
    }
    _getErrorMessage(exceptionDetails) {
        var _a, _b;
        if (exceptionDetails.text === UNCAUGHT_JS_ERROR)
            return (_a = exceptionDetails.exception) === null || _a === void 0 ? void 0 : _a.description;
        if (exceptionDetails.text === UNCAUGHT_PROMISE_REJECTION)
            return ((_b = exceptionDetails.exception) === null || _b === void 0 ? void 0 : _b.value) + `\n    ${NO_STACK_TRACE_AVAILABLE_MESSAGE}`;
        return void 0;
    }
    _getErrFilter(exceptionDetails, errorMessage) {
        var _a, _b, _c, _d, _e, _f;
        const stack = (_c = (_b = (_a = exceptionDetails.exception) === null || _a === void 0 ? void 0 : _a.preview) === null || _b === void 0 ? void 0 : _b.properties[0]) === null || _c === void 0 ? void 0 : _c.value;
        const message = (_f = (_e = (_d = exceptionDetails.exception) === null || _d === void 0 ? void 0 : _d.preview) === null || _e === void 0 ? void 0 : _e.properties[1]) === null || _f === void 0 ? void 0 : _f.value;
        return {
            stack: stack || errorMessage,
            pageUrl: exceptionDetails.url,
            message: message || errorMessage,
        };
    }
    _getShouldSkipJsErrors(errFilter) {
        const options = this.command ? this.command.options : this._testRun.prepareSkipJsErrorsOption();
        if (options === true)
            return true;
        if (typeof options === 'function')
            // @ts-ignore
            return options(errFilter);
        return false;
    }
    _addTestRunEvents() {
        this._testRun.once('done', this._dispose.bind(this));
    }
    _dispose() {
        this.command = null;
    }
}
exports.default = JSErrorHandlingAPI;
module.exports = exports.default;
