// testcafe/lib/proxyless/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_pipeline_1 = __importDefault(require("./request-pipeline"));
const add_custom_debug_formatters_1 = __importDefault(require("./add-custom-debug-formatters"));
const debug_loggers_1 = require("../utils/debug-loggers");
const console_messages_1 = __importDefault(require("./console-messages"));
const native_dialogs_1 = __importDefault(require("./native-dialogs"));
const js_error_handling_1 = __importDefault(require("./js-error-handling"));
const session_storage_1 = __importDefault(require("./session-storage"));
const ALL_REQUEST_RESPONSES = { requestStage: 'Request' };
const ALL_REQUEST_REQUESTS = { requestStage: 'Response' };
const ALL_REQUESTS_DATA = [ALL_REQUEST_REQUESTS, ALL_REQUEST_RESPONSES];
class Proxyless {
    constructor(browserId, client) {
        this._client = client;
        this.requestPipeline = new request_pipeline_1.default(browserId, client);
        this.consoleMessagesApi = new console_messages_1.default(browserId, client);
        this.nativeDialogsApi = new native_dialogs_1.default(browserId, client);
        this.sessionStorage = new session_storage_1.default(browserId, client);
        this.jsErrorHandlingApi = new js_error_handling_1.default(browserId, client);
        this.sessionStorage.on('contextStorageModified', sessionStorage => {
            const COMMAND_EXECUTING_FLAG = 'testcafe|driver|command-executing-flag';
            if (sessionStorage) {
                const parsed = JSON.parse(sessionStorage);
                if (parsed.hasOwnProperty(COMMAND_EXECUTING_FLAG)) {
                    const storage = {};
                    storage[COMMAND_EXECUTING_FLAG] = parsed[COMMAND_EXECUTING_FLAG];
                    this.requestPipeline.contextStorage = JSON.stringify(storage);
                }
            }
        });
        (0, add_custom_debug_formatters_1.default)();
    }
    async init(options) {
        // NOTE: We are forced to handle all requests and responses at once
        // because CDP API does not allow specifying request filtering behavior for different handlers.
        await this._client.Fetch.enable({
            patterns: ALL_REQUESTS_DATA,
        });
        const proxylessSystems = [
            this.requestPipeline,
            this.consoleMessagesApi,
            this.nativeDialogsApi,
            this.sessionStorage,
            this.jsErrorHandlingApi,
        ];
        for (const api of proxylessSystems)
            await api.init(options);
        (0, debug_loggers_1.proxylessLogger)('proxyless initialized');
    }
    async dispose() {
        this.requestPipeline.stop();
        await this._client.Fetch.disable();
        (0, debug_loggers_1.proxylessLogger)('proxyless disposed');
    }
}
exports.default = Proxyless;
module.exports = exports.default;
